source/extensions/transport_sockets/starttls/config.cc

```diff
@@ -36,17 +36,52 @@ Network::TransportSocketFactoryPtr DownstreamStartTlsSocketFactory::createTransp
       tls_socket_config_factory.createTransportSocketFactory(outer_config.tls_socket_config(),
                                                              context, server_names);
 
-  return std::make_unique<ServerStartTlsSocketFactory>(outer_config, std::move(raw_socket_factory),
-                                                       std::move(tls_socket_factory));
+  return std::make_unique<StartTlsSocketFactory>(std::move(raw_socket_factory),
+                                                 std::move(tls_socket_factory));
+}
+
+Network::TransportSocketFactoryPtr UpstreamStartTlsSocketFactory::createTransportSocketFactory(
+    const Protobuf::Message& message,
+    Server::Configuration::TransportSocketFactoryContext& context) {
+
+  const auto& outer_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::transport_sockets::starttls::v3::UpstreamStartTlsConfig&>(
+      message, context.messageValidationVisitor());
+
+  auto& raw_socket_config_factory = Config::Utility::getAndCheckFactoryByName<
+      Server::Configuration::UpstreamTransportSocketConfigFactory>(
+      TransportSocketNames::get().RawBuffer);
+
+  auto& tls_socket_config_factory = Config::Utility::getAndCheckFactoryByName<
+      Server::Configuration::UpstreamTransportSocketConfigFactory>(TransportSocketNames::get().Tls);
+
+  Network::TransportSocketFactoryPtr raw_socket_factory =
+      raw_socket_config_factory.createTransportSocketFactory(outer_config.cleartext_socket_config(),
+                                                             context);
+
+  Network::TransportSocketFactoryPtr tls_socket_factory =
+      tls_socket_config_factory.createTransportSocketFactory(outer_config.tls_socket_config(),
+                                                             context);
+
+  return std::make_unique<StartTlsSocketFactory>(std::move(raw_socket_factory),
+                                                 std::move(tls_socket_factory));
```

Comment:
can they share same implementation? isn't client connection starttls should emit `STARTTLS` to upstream and start handshake? I don't think that logic is in existing impelmentation.

Reply:
The `StartTlsSocket` is pretty basic in that it just forwards to the underlying raw transport until the `startSecureTransport()` method is called, and then switches over to the tls socket. If the protocol in question wants to emit a cleartext message to the tune of `STARTTLS` it is free to do so before calling `startSecureTransport()`. This is what happens in the postgres transport, albeit on the server side.

In the case of an upstream socket this should initiate the tls handshake as if it had stated life as tls. This works as expected in some custom code we're using internally and it works via the changes to the integration tests below.

Reply:
+1 for reusing the code. Config types are different, but the underlying structures are the same.

Reply:
This is about as much as I could share. The differences in argument lists for [Upstream|Downstream]TransportSocketFactory.createTransportSocketFactory makes it tough to do much more.

Reply:
> The `StartTlsSocket` is pretty basic in that it just forwards to the underlying raw transport until the `startSecureTransport()` method is called, and then switches over to the tls socket. If the protocol in question wants to emit a cleartext message to the tune of `STARTTLS` it is free to do so before calling `startSecureTransport()`. This is what happens in the postgres transport, albeit on the server side.

Can we document this behavior (expecting the network filter to emit `STARTTLS`) in the code? We need to make it clear to who will write a network filter.

Reply:
I've elaborated on this in the `TransportSocket` declaration.